router/usecase/pools/routable_cw_orderbook_pool.go

```diff
@@ -0,0 +1,246 @@
+	tokenInAfterTakerFee, _ := poolmanager.CalcTakerFeeExactIn(tokenIn, r.GetTakerFee())
+	return tokenInAfterTakerFee
+}
+
+// GetTakerFee implements domain.RoutablePool.
+func (r *routableOrderbookPoolImpl) GetTakerFee() math.LegacyDec {
+	return r.TakerFee
+}
+
+// SetTokenOutDenom implements domain.RoutablePool.
+func (r *routableOrderbookPoolImpl) SetTokenOutDenom(tokenOutDenom string) {
+	r.TokenOutDenom = tokenOutDenom
+}
+
+// CalcSpotPrice implements domain.RoutablePool.
+func (r *routableOrderbookPoolImpl) CalcSpotPrice(ctx context.Context, baseDenom string, quoteDenom string) (osmomath.BigDec, error) {
+	// Get the expected order direction
+	direction, err := r.GetDirection(quoteDenom, baseDenom)
+	if err != nil {
+		return osmomath.BigDec{}, err
+	}
+
+	tickIdx, err := r.GetStartTickIndex(direction.Opposite())
+	if err != nil {
+		return osmomath.BigDec{}, err
+	}
+
+	tick := r.OrderbookData.Ticks[tickIdx]
+
+	// Calculate the price for the current tick
+	return clmath.TickToPrice(tick.TickId)
+}
+
+// IsGeneralizedCosmWasmPool implements domain.RoutablePool.
+func (*routableOrderbookPoolImpl) IsGeneralizedCosmWasmPool() bool {
+	return false
+}
+
+// GetCodeID implements domain.RoutablePool.
+func (r *routableOrderbookPoolImpl) GetCodeID() uint64 {
+	return r.ChainPool.CodeId
+}
+
+// Determines order direction for the current orderbook given token in and out denoms
+// Returns:
+// - cosmwasmpool.BID (1) if the order is a bid (buying token out)
+// - cosmwasmpool.ASK (-1) if the order is an ask (selling token out)
+// - 0 if the order is not valid
+func (r *routableOrderbookPoolImpl) GetDirection(tokenInDenom, tokenOutDenom string) (domain.OrderbookDirection, error) {
+	if tokenInDenom == r.OrderbookData.BaseDenom && tokenOutDenom == r.OrderbookData.QuoteDenom {
+		return domain.ASK, nil
+	} else if tokenInDenom == r.OrderbookData.QuoteDenom && tokenOutDenom == r.OrderbookData.BaseDenom {
+		return domain.BID, nil
+	} else {
+		return 0, domain.OrderbookPoolMismatchError{PoolId: r.GetId(), TokenInDenom: tokenInDenom, TokenOutDenom: tokenOutDenom}
+	}
+}
+
+// Get the index for the tick state array for the starting index given direction
+func (r *routableOrderbookPoolImpl) GetStartTickIndex(direction domain.OrderbookDirection) (int, error) {
+	switch direction {
+	case domain.ASK:
+		return r.OrderbookData.NextAskTickIndex, nil
+	case domain.BID:
+		return r.OrderbookData.NextBidTickIndex, nil
+	default:
+		return -1, domain.OrderbookPoolInvalidDirectionError{Direction: direction}
+	}
+}
+
+// Converts an amount of token in to the value of token out given a price and direction
+func convertValue(amount osmomath.BigDec, price osmomath.BigDec, direction domain.OrderbookDirection) osmomath.BigDec {
+	switch direction {
+	case domain.ASK:
+		return amount.Mul(price)
+	case domain.BID:
+		return amount.Quo(price)
+	default:
+		return osmomath.ZeroBigDec()
+	}
```

Comment:
Having BID and ASK be booleans would remove the need for handling this default case as well as many other helper default cases down below.